Use Source Code Generator for PSVersionInfo class

iSazonov opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

The custom source generator allows us to exclude reflection and to directly get current PowerShell version from MSBuild.

PerfView doesn't work well on current .Net Preview. A fix has just been merged in .Net Runtime repo for this. I hope we get it in .Net 6.0 Preview6 or Preview7. So today I can not share reliable result from PerfView to confirm perf win. Obviously using static string is more faster then reflection.

## PR Context

Contribute to #13540
Related #14268

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

iSazonov committed:
Use Source Code Generator for PSVersionInfo class

iSazonov committed:
Fix CodeFactor issues

iSazonov committed:
Remove starting 'v' from _rawGitCommitId

iSazonov committed:
Fix spelling

iSazonov commented:
It seems test fails are not related to the PR - pwsh.exe has Preview.**8** in product version instead of Preview.**7**.

@rjmholt @TravisEz13 Could you please look the issue?

rjmholt commented:
@iSazonov opened https://github.com/PowerShell/PowerShell/issues/15607

rkeithhill commented on src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:
```diff
@@ -0,0 +1,84 @@
+            var result = CreatePSVersionInfoPartialClass(context);
+
+            // We must use specific file name suffix (*.g.cs,*.g, *.i.cs, *.generated.cs, *.designer.cs)
+            // so that Roslyng analizers skip the file.
```
Typo. Should be - `Roslyn analyzers`

iSazonov replied:
Thanks! 

Fixed.

iSazonov committed:
Fix typo

rkeithhill commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -42,6 +41,13 @@ public class PSVersionInfo
         private static readonly PSVersionHashTable s_psVersionTable;
 
+        /*
+            // The property and fields are generated by PSVersionInfoGenerator source generator.
```
Maybe tweak to `The following property ...`

iSazonov replied:
Fixed.

rkeithhill commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -42,6 +41,13 @@ public class PSVersionInfo
 
         private static readonly PSVersionHashTable s_psVersionTable;
 
+        /*
```
Use of multiline comment seems a bit out-of-place.  Consider using `//`

iSazonov replied:
Using `//` raise CodeFactor issue (blank line after the comment).

vexx32 replied:
I think CodeFactor recommends using `////` to comment out code specifically, which bypasses the usual warning for that.

iSazonov replied:
@xtqqczze Have you thoughts?

iSazonov commented on src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:
```diff
@@ -0,0 +1,84 @@
+    /// Source Code Generator to create partial PSVersionInfo class.
+    /// </summary>
+    [Generator]
+    public class PSVersionInfoGenerator : ISourceGenerator
```
Hmm, if I understand correctly the class is used in MSBuild and should be public but I did not check. In any case the class is not a part of PowerShell itself (it is not included in any PowerShell assembly).

iSazonov committed:
Fix comment

iSazonov committed:
Use ReleaseTag property

iSazonov commented:
@rjmholt I resolved the issue with release tag. It turned out that ci.psm1 always assigns ReleaseTag property directly to dotnet.

iSazonov commented on src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:
```diff
@@ -0,0 +1,86 @@
+        private static string CreatePSVersionInfoPartialClass(GeneratorExecutionContext context)
+        {
+            // We must put "<auto-generated" on first line so that Roslyng analizers skip the file.
+            const string SourceTemplate = @"
+// <auto-generated>
+// This file is auto-generated by PSVersionInfoGenerator.
+// </auto-generated>
+
+    namespace System.Management.Automation
+{{
+    public static partial class PSVersionInfo
+    {{
+        // Defined in PowerShell.Common.props as ProductVersion
+        // '6.0.0-beta.7 Commits: 29 SHA: 52c6b...'
+        internal static string ProductVersion {{ get; }} = ""{0}"";
+
+        // Defined in PowerShell.Common.props as
+        // git describe --abbrev=60 --long
+        // 'v6.0.0-beta.7-29-g52c6b...'
+        private static readonly string _rawGitCommitId = ""{1}"";
+
+        // Defined in PowerShell.Common.props as PSCoreBuildVersion
+        // '6.0.0-beta.7'
+        private static readonly string _mainVersion = ""{2}"";
+    }}
+}}";
```
Notice, the file is written to disk and it is accessible for compliance tools.

@PaulHigin Friendly ping.

PaulHigin replied:
@TravisEz13 , @iSazonov 
The concern, of course, is code injection.  I don't immediately see an issue, but this needs to be discussed in the next WG-Security meeting.  However, a meeting may not occur until December, due to the holidays.  Since this is an optional change, I feel it is not high priority.

PaulHigin replied:
To be clear, this is the opinion of the @WG-Security group, not just me.  We looked at `const string SourceTemplate` where the above property values are concatenated into the source template as strings.  This is a supply chain source code injection opportunity, that would be difficult to detect.  Our understanding is that using a CodeDom is the preferred way to build up source code for compilation.

iSazonov replied:
@PaulHigin I really don't understand about CodeDom. The best comment from .Net team I saw about using CodeDom in SGs is https://github.com/dotnet/roslyn/issues/48214#issuecomment-704432433

PaulHigin replied:
I'm not an expert either, and I'll defer to @TravisEz13.

SteveL-MSFT replied:
For code injection to happen here, there would need to be a double quote, right?  Can we just reject any strings that contain a double quote?

iSazonov replied:
I really don't understand. We could implement this as TypeGen and ResGen which we already use. Only difference is the TypeGen and ResGen are our custom utilities but the source generator is .Net feature.
To do injection, someone would have to compromise the entire repository in order to corrupt both build scripts and the source code.

daxian-dbw replied:
It's not clear if there is really a code injection vulnerability. The version properties are defined in `PowerShell.Common.props`, and then passed in by using the .NET source generator property `CompilerVisibleProperty`, so unless the properties are already tampered in `PowerShell.Common.props`, there seems no other way to cause a security problem, in my personal understanding.

Be noted that, today, the `ProductVersion` embedded in PowerShell assemblies is calculated using the same properties in `PowerShell.Common.props`. Then [in the static constructor](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/PSVersionInfo.cs#L77-L109) of `PSVersionInfo`, we get the `ProductVersion` from reflection, and parse the string to get PowerShell version and `GitCommitId`. So, if those properties could be tampered, then it's already a problem to our code today.

@TravisEz13 it would be very helpful if you can elaborate how the code injection could happen, and how would it affect our release builds or daily builds. Thanks!

daxian-dbw replied:
Comment resolved by https://github.com/PowerShell/PowerShell/pull/15603#issuecomment-1273819725

PaulHigin commented:
@iSazonov  Sorry for the delay.  I am unavailable today but will look at this early next week.

iSazonov committed:
Add s_psV72Version

daxian-dbw commented:
@iSazonov I really like you trying out the source generator!
I measured the code you removed from `PSVersionInfo.cs` by using the source generator, and it takes about `7ms` on average on my machine for private build (going through the `Contains(" Commit ")` code path, and hence slower). 

> ```
> Assembly currentAssembly = typeof(PSVersionInfo).Assembly;
> ProductVersion = currentAssembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion;
> 
> string rawGitCommitId;
> string mainVersion = ProductVersion.Substring(0, ProductVersion.IndexOf(' '));
> 
> if (ProductVersion.Contains(" Commits: "))
> {
>     rawGitCommitId = ProductVersion.Replace(" Commits: ", "-").Replace(" SHA: ", "-g");
> }
> else
> {
>     rawGitCommitId = mainVersion;
> }
> ```

The time it takes for `pwsh -noprofile -c exit` on my machine is about `390ms` on average (ranging from ~380 to ~420), and the time improvement is about `1.7%`.

Not sure if the gain worth the extra complexity though.

iSazonov commented:
> Not sure if the gain worth the extra complexity though.

It's no more complicated than our ResGen or TypeGen. It's just something new to get used to. 😄 

We could replace some of our Regex with a standard .Net Regex SG to get benefits in startup scenario and other scenarios.
And we could also replace TypeGen with our SG, which of course will not speed up PowerShell itself, but will noticeably speed up the CIs and the generated cache will probably be more correct.

daxian-dbw commented:
The current type initializer of PSVersionInfo takes `24-30 ms` on my dev machine, and it happens when creating the banner messgae, which is very early in the startup process. It would be good for the startup time if we can reduce the time it takes.

I will do an end-to-end measurement on the gains using source generator for `PSVersionInfo` plus the changes in https://github.com/PowerShell/PowerShell/pull/18210, and will report back the results.

daxian-dbw committed:
Merge part of #18210 to the source generator and also refactor it

daxian-dbw reviewed (commented):
I integrated the `FastParsePSVersion` change from #18210 to the source generator, and also refactored the code. After the change, the static constructor of `PSVersionTable` only takes `2-3 ms` on my dev machine (release build with R2R images, using the release tag `v7.3.0` to mimic a stable release). It was `24-30 ms` without the changes of this PR:

```
## Measured by wrapping the static constructor with `StopWatch`.

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5478
PowerShell 7.3.0
PS C:\> exit

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5993
PowerShell 7.3.0
PS C:\> exit

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5842
PowerShell 7.3.0
PS C:\> exit
```

daxian-dbw commented on src/System.Management.Automation/System.Management.Automation.csproj:
```diff
@@ -38,6 +56,11 @@
+    <!-- exclude code of source generators from compilation and IDEs (e.g. vscode and visual studio) -->
+    <Compile Remove="SourceGenerators\**\*.cs" />
+    <!-- exclude the generated source files because compiler already includes the in-memory source generator output -->
+    <Compile Remove="$(CompilerGeneratedFilesOutputPath)\**\*.cs" />
```
The only remaining problem is: the generated source code is not visible to IDEs like VSCode and Visual Studio.
The `<Compile Remove ... />` tag is needed because otherwise the compiler will report errors about duplicate declarations, as it automatically includes the im-memory source output, as well as the file generated from the last build.

In Visual Studio or VSCode, the constant members used in `PSVersionInfo.cs` cannot be resolved, because when using `<Compile Remove ... />`, the files will not be included for analyzing. I tried many ways, including attempting to delete the generated file using `<RemoveDir>` task in `System.Management.Automation.csproj` or `PSVersionInfoGenerator.csproj`, but nothing works. @iSazonov and @SeeminglyScience, really need help on this.

iSazonov replied:
Eh, it's been so long that I don't remember all the details anymore. Yes, there was such a problem and this line was a workaround. I even opened the case https://github.com/OmniSharp/omnisharp-vscode/issues/4215 and there's a progression in recent months, but I haven't checked if it's already in the current public version of VSCode/OmniSharp.
I think I saw in the VS 2022 announcement that it has SGs support, but I haven't checked and can't say for sure.

I'll try to find something useful

iSazonov replied:
I added cleanup commit:
- Turn off EmitCompilerGeneratedFiles - since latest VS Code works well we have no need to keep source of the generated file.
- Remove first blank line in generated code - cosmetic change

daxian-dbw replied:
It doesn't work in the following scenarios, for both VSCode and Visual Studio, which are quite common in development:
1. Run `Start-PSBuild` first, then open `System.Management.Automation` in VSCode
2. Open `System.Management.Automation` in VSCode to use an existing build, like from yesterday.

But I'm fine turning off `EmitCompilerGeneratedFiles` since it doesn't help. Also, I'm fine merging the PR without perfectly resolving this issue.

daxian-dbw replied:
I pushed a new commit to solve this issue, not perfectly, but works.

iSazonov replied:
I think this is fundamentally unsolvable because the code is generated on the fly and exists only at the moment of compilation. In other words OmniSharp must already be running to capture this code. 

> Also, I'm fine merging the PR without fully resolving this issue.

It's up to you. I don't mind if this workaround remains.

iSazonov committed:
Turn off EmitCompilerGeneratedFiles

- Turn off EmitCompilerGeneratedFiles
- Remove first blank line in generated code

iSazonov commented on src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:
```diff
@@ -0,0 +1,130 @@
+                result.preReleaseLabel);
+        }
+
+        private static (int major, int minor, int patch, string preReleaseLabel) FastParsePSVersion(string mainVersion)
+        {
+            // We only handle the pre-defined PSVersion format here, e.g. 7.x.x or 7.x.x-preview.x
```
We have no need to do "fast" on some ns since it is _build_ time.

Also, it is big headache to partially process versions in so many places (Build/Release.psm1, csproj, SG, SMA) - any minor change in one place can break all unpredictable.

_I strongly suggest keeping all version parsing in PowerShell.Common.props file._

As result, the SG will be very simple and stable, and our build process will be easily maintainable.

daxian-dbw replied:
> We have no need to do "fast" on some ns since it is build time.

It's just using the same function name. _fast_ is to compare with letting `SemanticVersion` do the string parsing.

> I strongly suggest keeping all version parsing in PowerShell.Common.props file.

I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

This change is essentially moving the version parsing code from `SMA` to source generator (`PSVersionInfo` itself doesn't do version parsing anymore). If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

iSazonov replied:
> I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

>If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

Exactly what is unlikely and this has already been concentrated in csproj for a long time. Now we are smearing it again.
When I opened this PR package tests were crashing because Release.psm1 was doing something incomprehensible. Now these tests are working, i.e. something was changed there again.
The release process is entirely on the MSFT team and above all you will waste your time over and over again. Although the community also releases their packages and they just benefit from having all the logic in csproj so they don't have to puzzle over the convoluted logic that is smeared out.

daxian-dbw replied:
> Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

1. It's not only about a few more seconds, it's more about not doing things that are not needed. Again, all other assemblies other than `S.M.A` don't need the version to be parsed.
2. Parsing the version in C# code is more maintainable and easier to change if really needed.

> The release process is entirely on the MSFT team and above all you will waste your time over and over again

I don't know about other maintainers' experience regarding this, but I really cannot remember what the last time was I saw a build failed due to parsing the version string in C# code.

iSazonov replied:
If we would think about third-party linux repository maintainer they have to build from sources for security reasons. And they have to learn and port Build.psm1 and other helpers again and again. They must also support previous versions (security patches). It is obvious how difficult this is. If we add that last year git tags for some reason have problems all the time, they have a big headache. It would be much easier if it just worked `dotnet publish`.

daxian-dbw replied:
There are too many `if's` here :)
As far as I understand, it's not a goal as of today to build PowerShell completely without scripts due to the complexity of the PowerShell release build.

But, even if things were changed in future, parsing version in source generator code would not be a blocker to that goal -- it would work well with just `dotnet publish`.

daxian-dbw committed:
Rename the function to be 'ParsePSVersion'

daxian-dbw committed:
Make generated files always visible to VSCode and Visual Studio and not affect re-building

daxian-dbw commented on src/System.Management.Automation/System.Management.Automation.csproj:
```diff
@@ -1,11 +1,29 @@
+    <!-- we persist source generator files under 'gen' folder so that they are visible to IDEs -->
+    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
+    <CompilerGeneratedFilesOutputPath>gen\SourceGenerated</CompilerGeneratedFilesOutputPath>
+  </PropertyGroup>
```
Generate the file, so VSCode and Visual Studio can always resolve the generated members.
Those files will be removed when building again, by `RemoveDir` in `PowerShell.Common.props`.

iSazonov replied:
I pushed a commit a day before (but you reverted now) and removed this old workaround because all works well in VS Code. We have no need to emit generated files.
Perhaps you need re-configure your local environment.
I use VS Code 1.72 and C# extension 1.25
![image](https://user-images.githubusercontent.com/22290914/194461998-943a41c5-b97a-4cc2-9e98-148fd160bbc3.png)

daxian-dbw replied:
You didn't see my reply to your comment? It's here: https://github.com/PowerShell/PowerShell/pull/15603#discussion_r989291958, and quoted below:

> It doesn't work in the following scenarios, for both VSCode and Visual Studio, which are quite common in development:
>1. Run `Start-PSBuild` first, then open `System.Management.Automation` in VSCode
> 2. Open `System.Management.Automation` in VSCode to use an existing build, like from yesterday.

iSazonov replied:
Sorry, too many comments already....

daxian-dbw commented on PowerShell.Common.props:
```diff
@@ -104,6 +94,9 @@
     </PropertyGroup>
 
+    <!-- Remove existing generated files by source generators because new source code will soon be generated when the build starts -->
+    <RemoveDir Directories="gen\SourceGenerated" Condition="Exists('gen\SourceGenerated')" />
```
The path `gen\SourceGenerated` is duplicated in here and `System.Management.Automation.csproj`. I tried creating a property holding the path in the `PropertyGroup` right above, and use it in `RemoveDir` and `CompilerGeneratedFilesOutputPath`. However, it works in `RemoveDir`, but not in `CompilerGeneratedFilesOutputPath`, and I cannot figure out why. I'd love if someone can make it work.

daxian-dbw commented:
@TravisEz13 and @PaulHigin, can you please review the changes again and re-evaluate the "code injection" concern?

Here is the doc (and the sample code) about the source generator: https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

The arguments used to format the source template in [PSVersionInfoGenerator.cs](https://github.com/PowerShell/PowerShell/pull/15603/files#diff-4757e7bd4af91fb154f76a6ecbb38da1ca557e1994cebfb674e355a64f6169e1) are from the build properties defined in `PowerShell.Common.props`.

daxian-dbw committed:
Merge branch 'master' into sg-psversion

iSazonov commented on PowerShell.Common.props:
```diff
@@ -73,16 +73,6 @@
       <PSCoreFormattedVersion Condition = "'$(ReleaseTag)' != '' or '$(PSCoreAdditionalCommits)' == '0'">$(PSCoreBuildVersion) SHA: $(PSCoreCommitSHA)</PSCoreFormattedVersion>
       <PSCoreFormattedVersion Condition = "'$(PSCoreFormattedVersion)' == ''">$(PSCoreBuildVersion) Commits: $(PSCoreAdditionalCommits) SHA: $(PSCoreCommitSHA)</PSCoreFormattedVersion>
 
-      <!-- Extract the major, minor and patch version numbers, as well as the preview label.
-           They are currently not used anywhere, so we comment them out for now.
-        <RegexSymVer>^((\d+).(\d+).(\d+))(?:-(.+))?</RegexSymVer>
-        <PSCorePrefixVersion>$([System.Text.RegularExpressions.Regex]::Match($(PSCoreBuildVersion), $(RegexSymVer)).Groups[1].Value)</PSCorePrefixVersion>
-        <PSCoreMajorVersion>$([System.Text.RegularExpressions.Regex]::Match($(PSCoreBuildVersion), $(RegexSymVer)).Groups[2].Value)</PSCoreMajorVersion>
-        <PSCoreMinorVersion>$([System.Text.RegularExpressions.Regex]::Match($(PSCoreBuildVersion), $(RegexSymVer)).Groups[3].Value)</PSCoreMinorVersion>
-        <PSCorePatchVersion>$([System.Text.RegularExpressions.Regex]::Match($(PSCoreBuildVersion), $(RegexSymVer)).Groups[4].Value)</PSCorePatchVersion>
-        <PSCoreLabelVersion>$([System.Text.RegularExpressions.Regex]::Match($(PSCoreBuildVersion), $(RegexSymVer)).Groups[5].Value)</PSCoreLabelVersion>
```
I wouldn't delete it. Might come in handy.

daxian-dbw replied:
Will bring them back.

iSazonov commented on src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:
```diff
@@ -0,0 +1,130 @@
+
+            var result = ParsePSVersion(mainVersion);
+
+            return string.Format(
+                CultureInfo.InvariantCulture,
+                SourceTemplate,
+                productVersion,
+                gitCommitId,
+                result.major,
+                result.minor,
+                result.patch,
+                result.preReleaseLabel);
```
Maybe use `string.Create()` and interpolated string. This will make the template more readable.

daxian-dbw replied:
I don't think `string.Create` is available in `netstandard2.0`.
Also, IMO, it's pretty readable already.

iSazonov commented on src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj:
```diff
@@ -0,0 +1,21 @@
+  <PropertyGroup>
+    <!-- source generator project needs to target 'netstandard2.0' -->
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <LangVersion>9.0</LangVersion>
```
```suggestion
    <LangVersion>10.0</LangVersion>
```

iSazonov commented on src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj:
```diff
@@ -0,0 +1,21 @@
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.3.1" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
```
It is needed only for publishing SG.
https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets
```suggestion
```

daxian-dbw replied:
Good point. Updated.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -42,6 +41,18 @@ public static class PSVersionInfo
+                internal const string ProductVersion;
+                internal const string GitCommitId;
+
+                private const int Version_Major
+                private const int Version_Minor;
+                private const int Version_Patch;
+                private const string Version_Label;
```
Are the names follow common naming standard?
Maybe MajorVersion and so on?

daxian-dbw replied:
What I intend to do is to make it look like `Version.Major`, `Version.Minor` and so on, to make their meaning more obvious.

As for naming standard, I did a quick search in PowerShell code base, and we do have const names using `_` in them, although most of them are const values for native code.

Personally, I think this is fine, especially that CodeFactor doesn't bark about it. But I'm willing to change if more people show concerns about the naming.

iSazonov replied:
It is minor. I am fine to merge as-is.

daxian-dbw committed:
Address comments

iSazonov reviewed (commented):
LGTM.

PaulHigin commented:
@WG-Security
The security group is Ok with this, because for release builds the generated version comes from a trusted source (releaseTag).

daxian-dbw reviewed (approved):
LGTM

daxian-dbw committed:
Merge branch 'master' into sg-psversion

daxian-dbw commented:
@iSazonov Just FYI, I re-did the conflict resolution because you accidentally brough back the `PSCompatibleVersions` property, which was deleted in the other PR.

iSazonov commented:
@daxian-dbw Thanks for working on the SG!.

iSazonov commented:
> @iSazonov Just FYI, I re-did the conflict resolution because you accidentally brough back the `PSCompatibleVersions` property, which was deleted in the other PR.

Gihub did show only one merge conflict for
```
       internal static string ProductVersion { get; }
        internal static string GitCommitId
        {
            get
            {
                return (string)s_psVersionTable[PSGitCommitIdName];
            }
        }
```
but did not marked  PSCompatibleVersions. I wonder why. 😕

daxian-dbw commented:
> Since you mentioned saving build time I think you could be interesting in SG for TypeGen.

I don't know if that would save build time, I doubt it :)
